File: xkey/pkg/nativemsg/crypto.go
// Copyright (c) 2025 Jeremy Hahn
// Copyright (c) 2025 Automate The Things, LLC
//
// This file is part of go-xkms.
//
// go-xkms is dual-licensed:
//
// 1. GNU Affero General Public License v3.0 (AGPL-3.0)
//    See LICENSE file or visit https://www.gnu.org/licenses/agpl-3.0.html
//
// 2. Commercial License
//    Contact licensing@automatethethings.com for commercial licensing options.

package nativemsg

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/rand"
	"encoding/binary"
	"io"
	"sync/atomic"

	"golang.org/x/crypto/hkdf"

	"crypto/sha256"
)

const (
	// SessionKeySize is the size of the AES-256-GCM key in bytes.
	SessionKeySize = 32

	// NonceSize is the size of the AES-256-GCM nonce in bytes.
	NonceSize = 12

	// HKDFInfo is the context string for HKDF key derivation.
	HKDFInfo = "xkey-nativemsg-v1"

	// X25519KeySize is the size of an X25519 public key.
	X25519KeySize = 32

	// hkdfInfoSend is the HKDF info string for the lower-pubkey peer's send key.
	hkdfInfoSend = HKDFInfo + "-send"

	// hkdfInfoRecv is the HKDF info string for the lower-pubkey peer's recv key.
	hkdfInfoRecv = HKDFInfo + "-recv"
)

// SessionCrypto handles per-connection X25519 ECDH key exchange and
// AES-256-GCM message encryption with replay protection.
//
// After construction via NewSessionCrypto, the caller must exchange public
// keys with the peer and call CompleteHandshake before Encrypt/Decrypt
// become available.
type SessionCrypto struct {
	privKey   *ecdh.PrivateKey
	pubKey    []byte
	sendKey   [SessionKeySize]byte
	recvKey   [SessionKeySize]byte
	sendNonce atomic.Uint64
	recvNonce atomic.Uint64
	ready     atomic.Bool
}

// NewSessionCrypto generates a new ephemeral X25519 keypair and returns
// the SessionCrypto instance along with the raw public key bytes that
// must be sent to the peer during the handshake.
func NewSessionCrypto() (*SessionCrypto, []byte, error) {
	curve := ecdh.X25519()
	privKey, err := curve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, ErrHandshakeFailed
	}
	pubBytes := privKey.PublicKey().Bytes()
	sc := &SessionCrypto{
		privKey: privKey,
		pubKey:  pubBytes,
	}
	return sc, pubBytes, nil
}

// CompleteHandshake performs the ECDH computation with the peer's public
// key and derives directional send/recv session keys via HKDF-SHA256.
//
// Key direction is determined by lexicographic comparison of the two
// public keys. The peer whose public key is lexicographically lower
// uses HKDF(shared, "xkey-nativemsg-v1-send") as its sendKey and
// HKDF(shared, "xkey-nativemsg-v1-recv") as its recvKey. The other
// peer reverses these assignments. This guarantees both sides derive
// identical keying material with opposite send/recv roles.
//
// CompleteHandshake may only be called once. Subsequent calls return
// ErrHandshakeFailed.
func (sc *SessionCrypto) CompleteHandshake(peerPubKey []byte) error {
	if sc.ready.Load() {
		return ErrHandshakeFailed
	}
	if len(peerPubKey) != X25519KeySize {
		return ErrInvalidPublicKey
	}
	curve := ecdh.X25519()
	peerKey, err := curve.NewPublicKey(peerPubKey)
	if err != nil {
		return ErrInvalidPublicKey
	}
	shared, err := sc.privKey.ECDH(peerKey)
	if err != nil {
		return ErrHandshakeFailed
	}

	// Derive two directional keys from the shared secret.
	keySend, err := deriveKey(shared, hkdfInfoSend)
	if err != nil {
		return ErrHandshakeFailed
	}
	keyRecv, err := deriveKey(shared, hkdfInfoRecv)
	if err != nil {
		return ErrHandshakeFailed
	}

	// The peer with the lexicographically lower public key uses
	// keySend as its sendKey and keyRecv as its recvKey. The other
	// peer swaps them so that each side's send corresponds to the
	// other's recv.
	if bytes.Compare(sc.pubKey, peerPubKey) <= 0 {
		sc.sendKey = keySend
		sc.recvKey = keyRecv
	} else {
		sc.sendKey = keyRecv
		sc.recvKey = keySend
	}

	sc.ready.Store(true)
	return nil
}

// Encrypt encrypts plaintext using AES-256-GCM with the send key and a
// monotonically increasing nonce counter. The returned nonce value and
// ciphertext must both be transmitted to the peer.
//
// The 12-byte GCM nonce is constructed as: 8-byte little-endian counter
// followed by 4 zero bytes.
func (sc *SessionCrypto) Encrypt(plaintext []byte) (nonce uint64, ciphertext []byte, err error) {
	if !sc.ready.Load() {
		return 0, nil, ErrHandshakeRequired
	}
	block, err := aes.NewCipher(sc.sendKey[:])
	if err != nil {
		return 0, nil, ErrEncryptionFailed
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return 0, nil, ErrEncryptionFailed
	}

	// Increment first so nonces start at 1; 0 is reserved as "no message
	// received yet" on the receiver side.
	counter := sc.sendNonce.Add(1)
	nonceBytes := counterToNonce(counter)

	sealed := gcm.Seal(nil, nonceBytes[:], plaintext, nil)
	return counter, sealed, nil
}

// Decrypt decrypts ciphertext using AES-256-GCM with the recv key.
// Messages with a nonce counter less than or equal to the last
// successfully-received nonce are rejected with ErrReplayDetected.
func (sc *SessionCrypto) Decrypt(nonce uint64, ciphertext []byte) ([]byte, error) {
	if !sc.ready.Load() {
		return nil, ErrHandshakeRequired
	}

	// Replay protection: reject nonces that have already been seen.
	for {
		last := sc.recvNonce.Load()
		if nonce <= last {
			return nil, ErrReplayDetected
		}
		if sc.recvNonce.CompareAndSwap(last, nonce) {
			break
		}
	}

	block, err := aes.NewCipher(sc.recvKey[:])
	if err != nil {
		return nil, ErrDecryptionFailed
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, ErrDecryptionFailed
	}

	nonceBytes := counterToNonce(nonce)
	plaintext, err := gcm.Open(nil, nonceBytes[:], ciphertext, nil)
	if err != nil {
		return nil, ErrDecryptionFailed
	}
	return plaintext, nil
}

// PublicKey returns the raw X25519 public key bytes for this session.
func (sc *SessionCrypto) PublicKey() []byte {
	out := make([]byte, len(sc.pubKey))
	copy(out, sc.pubKey)
	return out
}

// Ready returns true after CompleteHandshake has successfully derived
// the session keys.
func (sc *SessionCrypto) Ready() bool {
	return sc.ready.Load()
}

// deriveKey uses HKDF-SHA256 to derive a 32-byte key from the shared
// secret and an info string.
func deriveKey(shared []byte, info string) ([SessionKeySize]byte, error) {
	var key [SessionKeySize]byte
	r := hkdf.New(sha256.New, shared, nil, []byte(info))
	if _, err := io.ReadFull(r, key[:]); err != nil {
		return key, err
	}
	return key, nil
}

// counterToNonce constructs a 12-byte GCM nonce from an 8-byte
// little-endian counter followed by 4 zero bytes.
func counterToNonce(counter uint64) [NonceSize]byte {
	var nonce [NonceSize]byte
	binary.LittleEndian.PutUint64(nonce[:8], counter)
	return nonce
}

File: xkey/extension/src/crypto.ts
// xKey AutoFill Extension - Session Encryption
//
// Implements X25519 ECDH key agreement + HKDF-SHA256 key derivation +
// AES-256-GCM authenticated encryption using the Web Crypto API.
//
// Protocol:
//   1. Both sides generate ephemeral X25519 keypairs.
//   2. Public keys are exchanged in the handshake.
//   3. ECDH shared secret is derived.
//   4. HKDF-SHA256 derives two directional AES-256-GCM keys:
//      - info="xkey-nativemsg-v1-send"
//      - info="xkey-nativemsg-v1-recv"
//   5. Direction is determined by lexicographic comparison of public keys.
//      The party with the lower public key uses "send" as their send key.
//   6. Nonces are monotonic 64-bit counters, encoded as 12-byte IVs
//      (8 bytes LE uint64 + 4 zero bytes).

const HKDF_SEND_INFO = 'xkey-nativemsg-v1-send';
const HKDF_RECV_INFO = 'xkey-nativemsg-v1-recv';

const encoder = new TextEncoder();
const decoder = new TextDecoder();

/**
 * Compare two Uint8Arrays lexicographically.
 * Returns negative if a < b, zero if equal, positive if a > b.
 */
function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i] !== b[i]) {
      return a[i] - b[i];
    }
  }
  return a.length - b.length;
}

/** Encode bytes to base64 string. */
function toBase64(bytes: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/** Decode base64 string to bytes. */
function fromBase64(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Build a 12-byte AES-GCM nonce from a uint64 counter.
 * Layout: [counter as 8-byte LE uint64] [4 zero bytes]
 */
function buildNonce(counter: number): Uint8Array {
  const buf = new ArrayBuffer(12);
  const view = new DataView(buf);
  // Write counter as little-endian uint64.
  // JavaScript numbers are safe up to 2^53-1, which is more than sufficient
  // for a session nonce counter.
  const lo = counter & 0xFFFFFFFF;
  const hi = Math.floor(counter / 0x100000000) & 0xFFFFFFFF;
  view.setUint32(0, lo, true);
  view.setUint32(4, hi, true);
  // Bytes 8-11 remain zero.
  return new Uint8Array(buf);
}

/**
 * Derive an AES-256-GCM key using HKDF-SHA256 from a shared secret.
 */
async function deriveKey(sharedSecret: ArrayBuffer, info: string): Promise<CryptoKey> {
  const hkdfKey = await crypto.subtle.importKey(
    'raw',
    sharedSecret,
    'HKDF',
    false,
    ['deriveKey'],
  );

  return crypto.subtle.deriveKey(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: new Uint8Array(32), // 32 zero bytes
      info: encoder.encode(info),
    },
    hkdfKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );
}

/**
 * SessionCrypto manages X25519 ECDH key exchange and AES-256-GCM
 * session encryption for the native messaging channel.
 *
 * Usage:
 *   const session = await SessionCrypto.create();
 *   // Send session.getPublicKeyBase64() to peer in handshake
 *   // Receive peer's public key
 *   await session.completeHandshake(peerPubKeyBase64);
 *   // Now encrypt/decrypt:
 *   const { nonce, ciphertext } = await session.encrypt('hello');
 *   const plaintext = await session.decrypt(nonce, ciphertext);
 */
export class SessionCrypto {
  private sendKey: CryptoKey | null = null;
  private recvKey: CryptoKey | null = null;
  // Nonces start at 1 to match the Go native host convention.
  // Go uses atomic.Uint64.Add(1) which pre-increments (first nonce = 1),
  // and its Decrypt rejects nonce <= last (last starts at 0, so nonce 0
  // is always rejected).
  private sendNonce: number = 1;
  private recvNonce: number = 1;
  private publicKeyBytes: Uint8Array;
  private privateKey: CryptoKey;
  private _ready: boolean = false;

  private constructor(privateKey: CryptoKey, publicKeyBytes: Uint8Array) {
    this.privateKey = privateKey;
    this.publicKeyBytes = publicKeyBytes;
  }

  /**
   * Factory method: generate an ephemeral X25519 keypair and return
   * a new SessionCrypto instance ready for handshake.
   */
  static async create(): Promise<SessionCrypto> {
    const keyPair = await crypto.subtle.generateKey(
      'X25519' as unknown as EcKeyGenParams,
      false,
      ['deriveBits'],
    ) as CryptoKeyPair;

    // Export the public key so we can send it and compare it.
    const pubRaw = await crypto.subtle.exportKey('raw', keyPair.publicKey);
    const pubBytes = new Uint8Array(pubRaw);

    return new SessionCrypto(keyPair.privateKey, pubBytes);
  }

  /**
   * Complete the handshake by performing ECDH with the peer's public key
   * and deriving directional session keys.
   *
   * After this call, encrypt() and decrypt() are available.
   */
  async completeHandshake(peerPubKeyBase64: string): Promise<void> {
    const peerPubBytes = fromBase64(peerPubKeyBase64);

    // Import peer's public key for ECDH.
    const peerPubKey = await crypto.subtle.importKey(
      'raw',
      peerPubBytes.buffer as ArrayBuffer,
      'X25519' as unknown as EcKeyImportParams,
      false,
      [],
    );

    // Derive the shared secret via ECDH.
    const sharedSecret = await crypto.subtle.deriveBits(
      { name: 'X25519', public: peerPubKey } as unknown as EcdhKeyDeriveParams,
      this.privateKey,
      256,
    );

    // Derive two directional keys.
    const keySend = await deriveKey(sharedSecret, HKDF_SEND_INFO);
    const keyRecv = await deriveKey(sharedSecret, HKDF_RECV_INFO);

    // Determine direction by lexicographic comparison of public keys.
    // The party with the lower public key uses the "send"-labeled key
    // as their send key. The party with the higher public key swaps.
    const cmp = compareBytes(this.publicKeyBytes, peerPubBytes);
    if (cmp < 0) {
      // Our key is lower: send = SEND label, recv = RECV label
      this.sendKey = keySend;
      this.recvKey = keyRecv;
    } else {
      // Our key is higher (or equal, which should not happen): swap
      this.sendKey = keyRecv;
      this.recvKey = keySend;
    }

    this.sendNonce = 1;
    this.recvNonce = 1;
    this._ready = true;
  }

  /**
   * Encrypt a plaintext string using AES-256-GCM with the next send nonce.
   * Returns the nonce value and the base64-encoded ciphertext.
   */
  async encrypt(plaintext: string): Promise<{ nonce: number; ciphertext: string }> {
    if (!this.sendKey) {
      throw new Error('SessionCrypto: handshake not completed');
    }

    const nonce = this.sendNonce;
    this.sendNonce++;

    const iv = buildNonce(nonce);
    const plaintextBytes = encoder.encode(plaintext);

    const ciphertextBuf = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: iv.buffer as ArrayBuffer },
      this.sendKey,
      plaintextBytes,
    );

    return {
      nonce,
      ciphertext: toBase64(new Uint8Array(ciphertextBuf)),
    };
  }

  /**
   * Decrypt a ciphertext received from the peer.
   *
   * Enforces monotonic nonce ordering for replay protection:
   * the received nonce must equal the expected receive counter.
   */
  async decrypt(nonce: number, ciphertextBase64: string): Promise<string> {
    if (!this.recvKey) {
      throw new Error('SessionCrypto: handshake not completed');
    }

    // Replay protection: nonces must arrive in strict order.
    if (nonce !== this.recvNonce) {
      throw new Error(
        `SessionCrypto: unexpected nonce ${nonce}, expected ${this.recvNonce}`,
      );
    }
    this.recvNonce++;

    const iv = buildNonce(nonce);
    const ciphertextBytes = fromBase64(ciphertextBase64);

    const plaintextBuf = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: iv.buffer as ArrayBuffer },
      this.recvKey,
      ciphertextBytes.buffer as ArrayBuffer,
    );

    return decoder.decode(plaintextBuf);
  }

  /** Get our public key as a base64 string (for the handshake message). */
  getPublicKeyBase64(): string {
    return toBase64(this.publicKeyBytes);
  }

  /** Whether the handshake is complete and encryption is available. */
  get ready(): boolean {
    return this._ready;
  }
}
